### PR TITLE
Enable INS provisioning via SBP fileio [DEVC-448]

### PR DESCRIPTION
**Pull request by silverjam**
# Summary
*[DEVC-448] SBP write needs to allow Duro inertial license / IMU data to be written*

# Detail
Enable INS provisioning by opening up the permissions for the `/persistent`, `/data`, and `/media` directories.  Give the `sbp_fileio_daemon` the ability to check against a list of allowed paths, add rudimentary unit tests for this functionality.

# Testing
Per test case https://swiftnavigation.testrail.io/index.php?/cases/view/2339:

> You should be able to read, and list from the following allowed directories with sbp fileio.
> 
>     /data (where the firmware upgrade will write)
>     /persistent
>     /persistent/ndb
>     /media
>     /media/mmcbk0p*
>     /media/sda*
> 

```
$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --list /data
('192.168.22.222', 55555)
.
..

$ ls /data
PiksiMulti-v1.5.14.bin  data.bin                foobar.txt

$ cat /data/foobar.txt
123

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --write /data/foobar.txt
('192.168.22.222', 55555)

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --list /data
('192.168.22.222', 55555)
.
..
foobar.txt

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --read /data/foobar.txt
('192.168.22.222', 55555)
bytearray(b'123\n')

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --list /persistent/
('192.168.22.222', 55555)
.
..
ndb
config.ini
licenses
imu_cal.json

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --read /persistent/config.ini
('192.168.22.222', 55555)
bytearray(b'[ntrip]\nenable=False\nurl=http://...\ngga_out_interval=10\n[ethernet]\nip_config_mode=Static\nip_address=192.168.22.222\nnetmask=255.255.255.0\ngateway=192.168.22.1\n[uart0]\nmode=RTCMv3 IN\n[uart1]\nmode=Disabled\n[tcp_server0]\nenabled_sbp_messages=\n[skylark]\nenable=False\n[solution]\noutput_every_n_obs=1\n[acquisition]\nglonass_acquisition_enabled=False\n\0\0\0\0....')

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --write /persistent/config.ini
('192.168.22.222', 55555)

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --read /persistent/config.ini
('192.168.22.222', 55555)
bytearray(b'[ntrip]\nenable=False\nurl=http://...\ngga_out_interval=10\n[ethernet]\nip_config_mode=Static\nip_address=192.168.22.222\nnetmask=255.255.255.0\ngateway=192.168.22.1\n[uart0]\nmode=RTCMv3 IN\n[uart1]\nmode=Disabled\n[tcp_server0]\nenabled_sbp_messages=\n[skylark]\nenable=False\n[solution]\noutput_every_n_obs=1\n[acquisition]\nglonass_acquisition_enabled=True\n')

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --list /persistent/licenses
('192.168.22.222', 55555)
.
..
smoothpose_license.json

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --read /persistent/licenses/smoothpose_license.json
('192.168.22.222', 55555)
bytearray(b'{"license" : "0x8bc...",\n"product" : "smoothpose"}')

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --write /persistent/licenses/smoothpose_license.json
('192.168.22.222', 55555)

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --read /persistent/licenses/smoothpose_license.json
('192.168.22.222', 55555)
bytearray(b'{"license" : "0x8bc...",\\n"product" : "smoothpose foobar"}\n')

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --list /persistent/ndb
('192.168.22.222', 55555)
.
..
lgf
almanac_wn
almanac
gnss_capb
utc_params
l2c_capb
ephemeris
iono

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --read /persistent/ndb/iono
('192.168.22.222', 55555)
bytearray(b'v1.6.19\x00\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\xb6\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xb6')


$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --delete /persistent/ndb/iono
('192.168.22.222', 55555)

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --list /persistent/ndb
('192.168.22.222', 55555)
.
..
lgf
almanac_wn
almanac
gnss_capb
utc_params
l2c_capb
ephemeris

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --list /media/mmcblk0p1 | grep foobar

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --write /media/mmcblk0p1/foobar.txt
('192.168.22.222', 55555)

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --list /media/mmcblk0p1 | grep foobar
foobar.txt

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --read /media/mmcblk0p1/foobar.txt
('192.168.22.222', 55555)
bytearray(b'123\n')
```

Per test case https://swiftnavigation.testrail.io/index.php?/cases/view/2340

> You should be barred from listing any directory not explicitly allowed in prior test case. Specifically, test the ability to list
> /
> /bin
> /usr
> /var
> /lib
> /etc
> 
> ...
> 
> Test the security sensitive files
> 
> /bin/upgrade_tool
> ...

```
$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --list /usr/bin
('192.168.22.222', 55555)

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --list /var/log
('192.168.22.222', 55555)

$ ./piksi_tools/fileio.py  --tcp --port $THE_PIKSI:55555 --read /usr/bin/upgrade_tool
('192.168.22.222', 55555)
bytearray(b'\x00\x00\x00\x00\x00\x00\x00
```

Corresponding log messages:
```
Sep 11 04:03:28 piksi local0.warn sbp_fileio_daemon[13847]: Received FILEIO_READ_DIR request for path (/usr/bin) outside base directory (/media/,/persistent/,/data/), ignoring...

Sep 11 04:01:30 piksi local0.warn sbp_fileio_daemon[13847]: Received FILEIO_READ request for path (/usr/bin/upgrade_tool) outside base directory (/media/,/persistent/,/data/), ignoring...
```

# Known issues
The fileio.py script will occasionally return a buffer padded with a large number of NULL bytes, this appears to be an existing issue with fileio though.

### CODE REVIEW

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v1.5.0-develop-2018083120-11-g02d42f5d
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=660572)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-11-g02d42f5d&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-11-g02d42f5d&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v1.5.0-develop-2018083120-11-g02d42f5d
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v1.5.0-develop-2018083120-11-g02d42f5d)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.5.0-develop-2018083120-11-g02d42f5d/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.5.0-develop-2018083120-11-g02d42f5d

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v1.5.0-develop-2018083120-14-gf14fe057
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=661431)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-14-gf14fe057&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-14-gf14fe057&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v1.5.0-develop-2018083120-14-gf14fe057
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v1.5.0-develop-2018083120-14-gf14fe057)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.5.0-develop-2018083120-14-gf14fe057/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.5.0-develop-2018083120-14-gf14fe057

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v1.5.0-develop-2018090719-15-gcf047664
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=663377)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018090719-15-gcf047664&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018090719-15-gcf047664&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v1.5.0-develop-2018090719-15-gcf047664
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v1.5.0-develop-2018090719-15-gcf047664)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.5.0-develop-2018090719-15-gcf047664/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.5.0-develop-2018090719-15-gcf047664

File: package/sbp_fileio_daemon/sbp_fileio_daemon/test/run_sbp_fileio_daemon_tests.cc [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  ASSERT_TRUE( path_validator_allowed_count(ctx) == 1 );
+
+  path_validator_allow_path(ctx, "/fake_persist/blah/");
+  ASSERT_TRUE( path_validator_check(ctx, "/fake_persist/blah/blahblah.txt") );
+
+  ASSERT_TRUE( path_validator_check(ctx, "fake_persist/blah/blahblah.txt") );
+
+  ASSERT_TRUE( path_validator_allowed_count(ctx) == 2 );
+
+	const char* base_paths = path_validator_base_paths(ctx);
```
Comment (benjaminaltieri) [inline package/sbp_fileio_daemon/sbp_fileio_daemon/test/run_sbp_fileio_daemon_tests.cc]: indentation

File: package/sbp_fileio_daemon/sbp_fileio_daemon.mk [hunk 2]
```diff
@@ -6,21 +6,50 @@
 
 SBP_FILEIO_DAEMON_VERSION = 0.1
 SBP_FILEIO_DAEMON_SITE = \
-  "${BR2_EXTERNAL_piksi_buildroot_PATH}/package/sbp_fileio_daemon/src"
+  "${BR2_EXTERNAL_piksi_buildroot_PATH}/package/sbp_fileio_daemon/sbp_fileio_daemon"
 SBP_FILEIO_DAEMON_SITE_METHOD = local
 SBP_FILEIO_DAEMON_DEPENDENCIES = libuv nanomsg_custom libsbp libpiksi
 
+ifeq ($(BR2_BUILD_TESTS),y)
+	SBP_FILEIO_DEPENDENCIES += gtest
```
Comment (benjaminaltieri) [inline package/sbp_fileio_daemon/sbp_fileio_daemon.mk]: yay tests!

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+  LIST_INIT(&(ctx->path_list));
+  ctx->allowed_count = 0;
+
+  return ctx;
+}
+
+void path_validator_destroy(path_validator_t **pctx)
+{
+  if (*pctx == NULL)
```
Comment (benjaminaltieri) [inline package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c]: null check on `pctx` before deref

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+  if (strlen(path) == 0) {
+
+    fprintf(stderr, "ERROR: directories allowed for SBP fileio must not be empty.\n");
+    exit(EXIT_FAILURE);
+  }
+
+  path_node_t *node = malloc(sizeof(path_node_t));
+  node->path[0] = '\0';
+
+  strncpy(node->path, path, sizeof(node->path));
```
Comment (benjaminaltieri) [inline package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c]: we should make sure we check lengths here, `strncpy` is known to do some funny stuff: https://stackoverflow.com/a/1258577

maybe even add a test to cover path overflow?

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v1.5.0-develop-2018090719-23-gc8b15dfe
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=669780)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018090719-23-gc8b15dfe&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018090719-23-gc8b15dfe&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v1.5.0-develop-2018090719-23-gc8b15dfe
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v1.5.0-develop-2018090719-23-gc8b15dfe)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.5.0-develop-2018090719-23-gc8b15dfe/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.5.0-develop-2018090719-23-gc8b15dfe

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c (hunk 4 above)
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c]: Added a test and removed strncpy

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c (hunk 3 above)
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c]: fixed

File: package/sbp_fileio_daemon/sbp_fileio_daemon.mk (hunk 2 above)
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon.mk]: 🥇 🥇 YAYAYAYAY 🥇 🥇 

party

File: package/sbp_fileio_daemon/sbp_fileio_daemon/test/run_sbp_fileio_daemon_tests.cc (hunk 1 above)
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/test/run_sbp_fileio_daemon_tests.cc]: fixed, have yet to figure out why my vim config randomly inserts tabs

File: package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c (hunk 4 above)
Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c]: Oops, didn't actually remove this strncpy...

Comment (silverjam) [reply package/sbp_fileio_daemon/sbp_fileio_daemon/src/path_validator.c]: Ok, removed it for real now, and added an overflow test for this function.

**Comment by nicolekelley** [comment]
@swiftnav-ram -- here are testing
